Rewrite exhaustiveness in one pass

Nadrieril opened the pull request:
This is at least my 4th attempt at this in as many years x) Previous attempts were all too complicated or too slow. But we're finally here!

The previous version of the exhaustiveness algorithm computed reachability for each arm then exhaustiveness of the whole match. Since each of these steps does roughly the same things, this rewrites the algorithm to do them all in one go. I also think this makes things much simpler.

I also rewrote the documentation of the algorithm in depth. Hopefully it's up-to-date and easier to follow now. Plz comment if anything's unclear.

r? @oli-obk I think you're one of the rare other people to understand the exhaustiveness algorithm?

cc @varkor I know you're not active anymore, but if you feel like having a look you might enjoy this :D

Fixes https://github.com/rust-lang/rust/issues/79307

Nadrieril committed:
Keep rows with guards in the matrix

Nadrieril committed:
Fully rework the algorithm and its explanation

Nadrieril committed:
Respect `split` invariants for `Opaque`s

Nadrieril committed:
Add some tests

Nadrieril committed:
Tweak `lower_pattern`

Nadrieril committed:
Store wildcard row in the matrix

Nadrieril committed:
Clarify the `Wildcard`/`Missing` situation

Nadrieril committed:
Cleanup span passing

Nadrieril commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit 297ed32e8481e2d76327bd57b695ab0321f6e9e5 with merge c0c929ffde3a8fd52c575102c13325a2c68fedad...
<!-- homu: {"type":"TryBuildStarted","head_sha":"297ed32e8481e2d76327bd57b695ab0321f6e9e5","merge_sha":"c0c929ffde3a8fd52c575102c13325a2c68fedad"} -->

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/6761916621/job/18377279620) [(plain)](https://github.com/rust-lang/rust/commit/297ed32e8481e2d76327bd57b695ab0321f6e9e5/checks/18377279620/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
GITHUB_ACTION=__run_7
GITHUB_ACTIONS=true
GITHUB_ACTION_REF=
GITHUB_ACTION_REPOSITORY=
GITHUB_ACTOR=Nadrieril
GITHUB_API_URL=https://api.github.com
GITHUB_BASE_REF=master
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_262e0ffa-11a2-4266-b250-7a05adf357e2
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_PATH=/home/runner/work/_temp/_github_workflow/event.json
GITHUB_GRAPHQL_URL=https://api.github.com/graphql
GITHUB_HEAD_REF=linear-pass-take-4
GITHUB_JOB=pr
GITHUB_PATH=/home/runner/work/_temp/_runner_file_commands/add_path_262e0ffa-11a2-4266-b250-7a05adf357e2
GITHUB_REF=refs/pull/117611/merge
GITHUB_REF_NAME=117611/merge
GITHUB_REF_PROTECTED=false
---
GITHUB_SERVER_URL=https://github.com
GITHUB_SHA=3498f862912c29c914398650faf28e7ce951b3ca
GITHUB_STATE=/home/runner/work/_temp/_runner_file_commands/save_state_262e0ffa-11a2-4266-b250-7a05adf357e2
GITHUB_STEP_SUMMARY=/home/runner/work/_temp/_runner_file_commands/step_summary_262e0ffa-11a2-4266-b250-7a05adf357e2
GITHUB_TRIGGERING_ACTOR=Nadrieril
GITHUB_WORKFLOW_REF=rust-lang/rust/.github/workflows/ci.yml@refs/pull/117611/merge
GITHUB_WORKFLOW_SHA=3498f862912c29c914398650faf28e7ce951b3ca
GITHUB_WORKSPACE=/home/runner/work/rust/rust
GOROOT_1_19_X64=/opt/hostedtoolcache/go/1.19.13/x64
---
 Documenting rustc_lint v0.0.0 (/checkout/compiler/rustc_lint)
 Documenting rustc_hir_analysis v0.0.0 (/checkout/compiler/rustc_hir_analysis)
 Documenting rustc_passes v0.0.0 (/checkout/compiler/rustc_passes)
 Documenting rustc_mir_build v0.0.0 (/checkout/compiler/rustc_mir_build)
error: unresolved link to `Matrix::push`
    |
    |
663 | /// Invariant: the first column must not contain or-patterns. This is handled by [`Matrix::push`].
    |                                                                                    ^^^^^^^^^^^^ the struct `Matrix` has no field or associated item named `push`
    |
    = note: `-D rustdoc::broken-intra-doc-links` implied by `-D warnings`
    = help: to override `-D warnings` add `#[allow(rustdoc::broken_intra_doc_links)]`
 Documenting rustc_const_eval v0.0.0 (/checkout/compiler/rustc_const_eval)
error: could not document `rustc_mir_build`

Caused by:
Caused by:
  process didn't exit successfully: `/checkout/obj/build/bootstrap/debug/rustdoc --edition=2021 --crate-type lib --crate-name rustc_mir_build compiler/rustc_mir_build/src/lib.rs --target x86_64-unknown-linux-gnu -o /checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/doc -Zunstable-options --check-cfg 'values(feature)' --check-cfg 'names()' --check-cfg 'values()' --error-format=json --json=diagnostic-rendered-ansi,artifacts,future-incompat -C metadata=d87e6a37983db245 -L dependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps -L dependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/release/deps --extern either=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/libeither-49c9c04d444dc07c.rmeta --extern rustc_apfloat=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_apfloat-d887aa7e85377076.rmeta --extern rustc_arena=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_arena-4cd536b6c13febce.rmeta --extern rustc_ast=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_ast-022bc10cc8c5b631.rmeta --extern rustc_data_structures=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_data_structures-62ec870da091c3dc.rmeta --extern rustc_errors=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_errors-1957ccbf9b46423e.rmeta --extern rustc_fluent_macro=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/release/deps/librustc_fluent_macro-10924ca5c421c8ab.so --extern rustc_hir=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_hir-c23923303847865a.rmeta --extern rustc_index=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_index-51083832d83671cc.rmeta --extern rustc_infer=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_infer-4d7d29e9bc2d4697.rmeta --extern rustc_macros=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/release/deps/librustc_macros-a3009b08c66bb671.so --extern rustc_middle=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_middle-188f4f864cc4c0f7.rmeta --extern rustc_serialize=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_serialize-fa42cbcbb78c13c4.rmeta --extern rustc_session=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_session-510e0881c7decc0d.rmeta --extern rustc_span=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_span-d25f7166e1949731.rmeta --extern rustc_target=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_target-5e247b22e59d568f.rmeta --extern rustc_trait_selection=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_trait_selection-50a42152fed44a64.rmeta --extern smallvec=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/libsmallvec-720d9ea4f4342760.rmeta --extern tracing=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/libtracing-0dc9638b9236e7df.rmeta --extern-html-root-url 'either=https://docs.rs/either/1.8.1/' --extern-html-root-url 'rustc_apfloat=https://docs.rs/rustc_apfloat/0.2.0+llvm-462a31f5a5ab/' --extern-html-root-url 'smallvec=https://docs.rs/smallvec/1.11.0/' --extern-html-root-url 'tracing=https://docs.rs/tracing/0.1.37/' -Zunstable-options --cfg=bootstrap --cfg=windows_raw_dylib -Csymbol-mangling-version=v0 -Zunstable-options '--check-cfg=values(bootstrap)' '--check-cfg=values(parallel_compiler)' '--check-cfg=values(no_btreemap_remove_entry)' '--check-cfg=values(crossbeam_loom)' '--check-cfg=values(span_locations)' '--check-cfg=values(rustix_use_libc)' '--check-cfg=values(emulate_second_only_system)' '--check-cfg=values(windows_raw_dylib)' --document-private-items --document-hidden-items -Dwarnings '-Wrustdoc::invalid_codeblock_attributes' --crate-version '1.75.0-nightly (3498f8629 2023-11-05)' --document-private-items '-Arustdoc::private-intra-doc-links' --enable-index-page -Zunstable-options -Znormalize-docs --show-type-layout --generate-link-to-definition '-Zcrate-attr=warn(rust_2018_idioms)' --extern-html-root-url 'ena=https://docs.rs/ena/latest/'` (exit status: 1)
Build completed unsuccessfully in 0:01:02
  local time: Sun Nov  5 14:42:23 UTC 2023
  network time: Sun, 05 Nov 2023 14:42:23 GMT
##[error]Process completed with exit code 1.
```

</details>

Nadrieril commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit ab918af40488cea0cf349aeeac00156181eac6a5 with merge 14c4d2ff18c46f09d9d0e40e316521523bd510b8...
<!-- homu: {"type":"TryBuildStarted","head_sha":"ab918af40488cea0cf349aeeac00156181eac6a5","merge_sha":"14c4d2ff18c46f09d9d0e40e316521523bd510b8"} -->

Nilstrieb commented:
you'll have to wait until january for oli to appear again :)

Nadrieril commented:
Ok! Thanks for the heads-up. Know anyone else familiar with exhaustiveness? @cjgillot how do you feel about it?

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/6762001592/job/18378293996)
Build commit: 14c4d2ff18c46f09d9d0e40e316521523bd510b8 (`14c4d2ff18c46f09d9d0e40e316521523bd510b8`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/6762001592/job/18378293996"},"merge_sha":"14c4d2ff18c46f09d9d0e40e316521523bd510b8"} -->

cjgillot commented:
I'll take a look during the week or next week-end.

Nadrieril commented:
Thank you!

varkor commented on compiler/rustc_mir_build/src/thir/pattern/usefulness.rs:
```diff
@@ -1,46 +1,55 @@
-//! I (Nadrieril) prefer to put new tests in `ui/pattern/usefulness` unless there's a specific
-//! reason not to, for example if they depend on a particular feature like `or_patterns`.
+//! The algorithm implemented here is inspired from the one described in [this
+//! paper](http://moscova.inria.fr/~maranget/papers/warn/index.html). We have however changed it in
```
Perhaps once this change is merged, it would be worth revisiting the idea of writing a paper about the pattern matching algorithm in rustc, given how much has changed from algorithm described in the original paper...

Nadrieril replied:
I was just thinking that the other day! Specifically because I'm working on never patterns atm which are a very unique pattern-matching feature. But you're right what we already have is pretty rad.

Full honesty though: I'd be more than happy to assist someone, but I'm not writing a paper again if I can help it x)

Nadrieril replied:
Would you like to write it? I'd have time to help you these days

varkor replied:
I'd be interested in working on one, but don't have the time at the moment. Perhaps some time in the new year?

Nadrieril replied:
Let's keep in touch :D

rust-timer commented:
Finished benchmarking commit (14c4d2ff18c46f09d9d0e40e316521523bd510b8): [comparison URL](https://perf.rust-lang.org/compare.html?start=04817ff00cfa1ec00a9babcb2ade1e3f2874d9bd&end=14c4d2ff18c46f09d9d0e40e316521523bd510b8&stat=instructions:u).

### Overall result: ❌✅ regressions and improvements - ACTION NEEDED


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please fix the regressions and do another perf run. If the next run shows neutral or positive results, the label will be automatically removed.

@bors rollup=never
@rustbot label: -S-waiting-on-perf +perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=04817ff00cfa1ec00a9babcb2ade1e3f2874d9bd&end=14c4d2ff18c46f09d9d0e40e316521523bd510b8&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 2.3%  | [2.2%, 2.4%]   | 6     |
| Improvements ✅ <br /> (primary)   | -1.0% | [-1.9%, -0.3%] | 10    |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -1.0% | [-1.9%, -0.3%] | 10    |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=04817ff00cfa1ec00a9babcb2ade1e3f2874d9bd&end=14c4d2ff18c46f09d9d0e40e316521523bd510b8&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.9%  | [0.6%, 1.2%]   | 4     |
| Regressions ❌ <br /> (secondary)  | 0.6%  | [0.6%, 0.6%]   | 1     |
| Improvements ✅ <br /> (primary)   | -1.2% | [-1.6%, -0.8%] | 2     |
| Improvements ✅ <br /> (secondary) | -5.0% | [-5.0%, -5.0%] | 1     |
| All ❌✅ (primary)                 | 0.2%  | [-1.6%, 1.2%]  | 6     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=04817ff00cfa1ec00a9babcb2ade1e3f2874d9bd&end=14c4d2ff18c46f09d9d0e40e316521523bd510b8&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -0.5% | [-0.5%, -0.5%] | 2     |
| Improvements ✅ <br /> (secondary) | -0.6% | [-0.6%, -0.6%] | 1     |
| All ❌✅ (primary)                 | -0.5% | [-0.5%, -0.5%] | 2     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=04817ff00cfa1ec00a9babcb2ade1e3f2874d9bd&end=14c4d2ff18c46f09d9d0e40e316521523bd510b8&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 638.392s -> 636.7s (-0.27%)
**Artifact size**: 304.54 MiB -> 304.56 MiB (0.01%)

Nadrieril commented:
We can ignore the match-stress benchmark, it shows no change on cycles or wall-time

Nadrieril commented:
Incidentally this fixes https://github.com/rust-lang/rust/issues/79307 :D

Nadrieril commented:
@cjgillot do you expect to have time for this anytime soon?

cjgillot commented on compiler/rustc_mir_build/src/thir/pattern/usefulness.rs:
```diff
@@ -471,45 +656,113 @@ impl<'p, 'tcx> fmt::Debug for PatStack<'p, 'tcx> {
     ) -> Matrix<'p, 'tcx> {
-        let mut matrix = Matrix::empty();
-        for row in &self.patterns {
+        let wildcard_row = self.wildcard_row.pop_head_constructor(pcx, ctor, usize::MAX);
```
Can we avoid using a magic `usize::MAX` value?

Nadrieril replied:
The issue is that I reuse `PatStack` for `wildcard_row` even though I don't need its row-tracking or reachability-tracking logic. You motivated me to split that off.

cjgillot reviewed (commented):
The high-level structure looks good. I was not able to review in-depth the implementation, but the test coverage is decent enough.

cjgillot commented:
> Incidentally this fixes #79307 :D

Is there a test for this? Can you add it to the OP for github auto-close?

Nadrieril commented on compiler/rustc_mir_build/src/thir/pattern/deconstruct_pat.rs:
```diff
@@ -617,6 +624,18 @@ impl Slice {
+    fn new() -> Self {
+        use std::sync::atomic::{AtomicU32, Ordering};
+        static OPAQUE_ID: AtomicU32 = AtomicU32::new(0);
+        OpaqueId(OPAQUE_ID.fetch_add(1, Ordering::SeqCst))
```
We need some way to recognize "this is the same Opaque" but we can't inspect the underlying `Const` value. On a previous iteration I used row ids to distinguish the opaques, but this breaks with or-patterns. So we need this id, and it has to be unique within a given match expression. The only alternative I see would be to track a unique id counter during lowering; I prefer this well-contained global state.

Nadrieril commented on compiler/rustc_mir_build/src/thir/pattern/deconstruct_pat.rs:
```diff
@@ -41,6 +41,13 @@
+//!
+//! ## Opaque patterns
+//!
+//! Some patterns, such as TODO, cannot be inspected, which we handle with `Constructor::Opaque`.
```
I fixed that in a later commit

Nadrieril committed:
Separate `PatStack` and `MatrixRow`

This disentangles the row-specific tracking of `parent_row` etc from the
logical operation of specialization. This means `wildcard_row` doesn't
need to provide dummy values for `parent_row` etc anymore.

Nadrieril commented on compiler/rustc_mir_build/src/thir/pattern/usefulness.rs:
```diff
@@ -384,21 +565,39 @@ impl<'a, 'p, 'tcx> fmt::Debug for PatCtxt<'a, 'p, 'tcx> {
+    /// specialized matrix. When we unspecialize, we use this to propagate reachability back up the
+    /// callstack.
+    /// At the start of the algorithm, this is the id of the arm this comes from (but we don't use
+    /// this fact anywhere).
```
Yeah, I did that

Nadrieril commented:
Thanks for taking the time to review this!

> Is there a test for this?

Yep, [here](https://github.com/rust-lang/rust/pull/117611/commits/bca172ad5b74c14815768ea43a17ba5a0c1770cb#diff-86cb91b509efdeef773165e479a79cc35da21fd4597e2af6692525630a93ef88R86-R90).

bors commented:
:umbrella: The latest upstream changes (presumably #118143) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

Nadrieril committed:
Abstract over the list of `WitnessStack`s

Nadrieril commented:
@rustbot ready

cjgillot commented:
@bors r+

bors commented:
:pushpin: Commit 273cbb73047436927a680da5d1636d87be9aafe3 has been approved by `cjgillot`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=cjgillot 273cbb73047436927a680da5d1636d87be9aafe3 -->
<!-- homu: {"type":"Approved","sha":"273cbb73047436927a680da5d1636d87be9aafe3","approver":"cjgillot","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 273cbb73047436927a680da5d1636d87be9aafe3 with merge ee80c8d0a8bc63b69f68216c5d37f9ab837eedd0...
<!-- homu: {"type":"BuildStarted","head_sha":"273cbb73047436927a680da5d1636d87be9aafe3","merge_sha":"ee80c8d0a8bc63b69f68216c5d37f9ab837eedd0"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/6992187837/job/19025012196)
Approved by: cjgillot
Pushing ee80c8d0a8bc63b69f68216c5d37f9ab837eedd0 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"cjgillot","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/6992187837/job/19025012196"},"merge_sha":"ee80c8d0a8bc63b69f68216c5d37f9ab837eedd0"} -->

rust-timer commented:
Finished benchmarking commit (ee80c8d0a8bc63b69f68216c5d37f9ab837eedd0): [comparison URL](https://perf.rust-lang.org/compare.html?start=f5dc2653fdd8b5d177b2ccbd84057954340a89fc&end=ee80c8d0a8bc63b69f68216c5d37f9ab837eedd0&stat=instructions:u).

### Overall result: ❌✅ regressions and improvements - ACTION NEEDED


**Next Steps**: If you can justify the regressions found in this perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please open an issue or create a new PR that fixes the regressions, add a comment linking to the newly created issue or PR, and then add the `perf-regression-triaged` label to this PR.

@rustbot label: +perf-regression
cc @rust-lang/wg-compiler-performance



### [Instruction count](https://perf.rust-lang.org/compare.html?start=f5dc2653fdd8b5d177b2ccbd84057954340a89fc&end=ee80c8d0a8bc63b69f68216c5d37f9ab837eedd0&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 1.1%  | [1.0%, 1.1%]   | 2     |
| Regressions ❌ <br /> (secondary)  | 1.6%  | [0.3%, 2.4%]   | 9     |
| Improvements ✅ <br /> (primary)   | -0.9% | [-2.0%, -0.2%] | 11    |
| Improvements ✅ <br /> (secondary) | -0.2% | [-0.2%, -0.2%] | 1     |
| All ❌✅ (primary)                 | -0.6% | [-2.0%, 1.1%]  | 13    |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=f5dc2653fdd8b5d177b2ccbd84057954340a89fc&end=ee80c8d0a8bc63b69f68216c5d37f9ab837eedd0&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 2.3%  | [0.8%, 4.8%]   | 3     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -1.1% | [-1.3%, -1.0%] | 4     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=f5dc2653fdd8b5d177b2ccbd84057954340a89fc&end=ee80c8d0a8bc63b69f68216c5d37f9ab837eedd0&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -0.9% | [-0.9%, -0.9%] | 1     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=f5dc2653fdd8b5d177b2ccbd84057954340a89fc&end=ee80c8d0a8bc63b69f68216c5d37f9ab837eedd0&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 674.609s -> 672.939s (-0.25%)
**Artifact size**: 313.33 MiB -> 313.35 MiB (0.01%)

nnethercote commented:
The icount regressions on secondary stress tests are balanced by improvement on primary benchmarks, which is fine.

@rustbot label: +perf-regression-triaged
